Logstash/logstash-input-azurewadtable/lib/logstash/inputs/azurewadtable.rb

```diff
@@ -63,99 +75,103 @@ def build_latent_query
-            toReplace = eventMessage.scan(/%\d+/)
-            payload = message.scan(/(?<!\\S)([a-zA-Z]+)=(\"[^\"]*\")(?!\\S)/)
-            # Split up the format string to seperate all of the numbers
-            toReplace.each do |key|
-              @logger.debug("Replacing key: " + key.to_s)
-              index = key.scan(/\d+/).join.to_i
-              newValue = payload[index - 1][1]
-              @logger.debug("New Value: " + newValue)
-              eventMessage[key] = newValue
-            end # do block
-            event.set("EventMessage", eventMessage)
-            @logger.debug("pretty print end. result: " + event.get("EventMessage").to_s)
-          end
-        end
-        decorate(event)
-        if event.get('PreciseTimeStamp').is_a?(Time)
-          event.set('PreciseTimeStamp', LogStash::Timestamp.new(event.get('PreciseTimeStamp')))
-        end
-        theTIMESTAMP = event.get('TIMESTAMP')
-        if theTIMESTAMP.is_a?(LogStash::Timestamp)
-          last_good_timestamp = theTIMESTAMP.to_iso8601
-        elsif theTIMESTAMP.is_a?(Time)
-          last_good_timestamp = theTIMESTAMP.iso8601
-          event.set('TIMESTAMP', LogStash::Timestamp.new(theTIMESTAMP))
-        else
-          @logger.warn("Found result with invalid TIMESTAMP. " + event.to_hash.to_s)
-        end
-        output_queue << event
-      end # each block
-      @idle_delay = 0
+
+    query = build_latent_query
+    filter_result = @duplicate_detector.filter_duplicates(query, ->(entity) {
+      last_good_timestamp = on_new_data(entity, output_queue, last_good_timestamp)
+    })
+
+    if filter_result
       if (!last_good_timestamp.nil?)
-        @last_timestamp = last_good_timestamp unless @continuation_token
+        @last_timestamp = last_good_timestamp
       end
     else
       @logger.debug("No new results found.")
-      @idle_delay = @idle_delay_seconds
-    end # if block
+    end
 
   rescue => e
-    @logger.error("Oh My, An error occurred.", :exception => e)
+    @logger.error("Oh My, An error occurred. Error:#{e}: Trace: #{e.backtrace}", :exception => e)
     raise
   end # process
 
+  def on_new_data(entity, output_queue, last_good_timestamp)
+    #@logger.debug("new event")
+    event = LogStash::Event.new(entity.properties)
+    event.set("type", @table_name)
+
+    # Help pretty print etw files
+    if (@etw_pretty_print && !event.get("EventMessage").nil? && !event.get("Message").nil?)
+      @logger.debug("event: " + event.to_s)
+      eventMessage = event.get("EventMessage").to_s
+      message = event.get("Message").to_s
+      @logger.debug("EventMessage: " + eventMessage)
+      @logger.debug("Message: " + message)
+      if (eventMessage.include? "%")
+        @logger.debug("starting pretty print")
+        toReplace = eventMessage.scan(/%\d+/)
+        payload = message.scan(/(?<!\\S)([a-zA-Z]+)=(\"[^\"]*\")(?!\\S)/)
+        # Split up the format string to seperate all of the numbers
```

Comment:
Can you add a comment here to describe what this processing is doing?
The code existed before, but I have to say, I still couldn't figure it after reading it 3 times.

Reply:
I have no idea what the pretty print does or for what kind of ETW traces it works.
I've tried pushing ETW data in the tables by following this walkthrough: https://docs.microsoft.com/en-us/azure/cloud-services/cloud-services-dotnet-diagnostics 
The data collected doesn't follow any of the expected rules in the plugin. I'll ask the original author, maybe he remembers

Reply:
ETW event message is, basically, a format string. This code tries to detect references to payload properties and replace them with payload property values.